File: app/src/main/java/dev/bee/kanjianki/backup/DatabaseBackupScheduler.java
package dev.bee.kanjianki.backup;

import android.content.Context;

import androidx.work.ExistingPeriodicWorkPolicy;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkManager;

import java.util.concurrent.TimeUnit;

public final class DatabaseBackupScheduler {
    private static final String UNIQUE_WORK_NAME = "kani_daily_db_backup";

    private DatabaseBackupScheduler() {
    }

    public static void schedule(Context context) {
        Context appContext = context.getApplicationContext();
        PeriodicWorkRequest request = new PeriodicWorkRequest.Builder(
                DatabaseBackupWorker.class,
                1,
                TimeUnit.DAYS,
                6,
                TimeUnit.HOURS
        ).build();
        WorkManager.getInstance(appContext).enqueueUniquePeriodicWork(
                UNIQUE_WORK_NAME,
                ExistingPeriodicWorkPolicy.KEEP,
                request
        );
    }

    public static void cancel(Context context) {
        cancel(context, appContext -> workName ->
                WorkManager.getInstance(appContext).cancelUniqueWork(workName));
    }

    static void cancel(Context context, WorkCancellerFactory factory) {
        factory.create(context.getApplicationContext()).cancelUniqueWork(UNIQUE_WORK_NAME);
    }

    interface WorkCancellerFactory {
        WorkCanceller create(Context appContext);
    }

    interface WorkCanceller {
        void cancelUniqueWork(String workName);
    }
}

File: app/src/main/kotlin/dev/bee/kanjianki/backup/DatabaseBackupScheduler.kt
package dev.bee.kanjianki.backup

import android.content.Context
import android.os.Build
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.PeriodicWorkRequest
import dev.bee.kanjianki.automation.AndroidWorkManagerGateway
import dev.bee.kanjianki.automation.WorkManagerGateway
import dev.bee.kanjianki.core.DatabaseBackupAvailabilityPolicy
import java.util.concurrent.TimeUnit

object DatabaseBackupScheduler {
    private const val UNIQUE_WORK_NAME = "kani_daily_db_backup"

    @JvmStatic
    fun schedule(context: Context?) {
        val appContext = context!!.applicationContext
        schedule(Build.VERSION.SDK_INT, AndroidWorkManagerGateway(appContext))
    }

    internal fun schedule(apiLevel: Int, backend: WorkManagerGateway) {
        if (!DatabaseBackupAvailabilityPolicy.forAndroidApi(apiLevel).operationsAllowed) {
            backend.cancelUniqueWork(UNIQUE_WORK_NAME)
            return
        }
        val request = PeriodicWorkRequest.Builder(
            DatabaseBackupWorker::class.java,
            1,
            TimeUnit.DAYS,
            6,
            TimeUnit.HOURS,
        ).build()
        backend.enqueueUniquePeriodicWork(
            UNIQUE_WORK_NAME,
            ExistingPeriodicWorkPolicy.KEEP,
            request,
        )
    }

    @JvmStatic
    fun cancel(context: Context?) {
        AndroidWorkManagerGateway(context!!.applicationContext).cancelUniqueWork(UNIQUE_WORK_NAME)
    }
}
